### ISSUE TITLE
Turbolinks 5 not working with Bootstrap-checkbox-radio-switch.js

### ISSUE THREAD

**Issue by fabriziobertoglio1987** [turbolinks/turbolinks#289]
Hello,

I fixed this by including in the html properties of the links the `'data-method'=>'get' ` or the `method: :get` for the `link_to` rails tags. The result was disabling turbolinks for that links..

I am not good in Javascript and Jquery. 

I don't know how to modify this js file so that it loads with turbolinks.. I do not see any `$(document).ready()` statement or similar.

The problem is connected to that file not loading when I change page with turbolinks. I set breakpoints and tested both scenario so I am sure.

This is the visual result, with turbolinks:

![screenshot from 2017-05-15 18-32-03](https://cloud.githubusercontent.com/assets/24992535/26069605/c9bd041a-39a1-11e7-99e4-4ae62f49a122.png)

Without turbolinks:

![screenshot from 2017-05-15 19-05-00](https://cloud.githubusercontent.com/assets/24992535/26069611/d3e96adc-39a1-11e7-8c44-d0e68aa5ebe6.png)

I attach the js file, but my question is mostly about this lines:

```
!function ($) {
}(window.jQuery);

!function ($) {
    var Radio = function (element, options) {
           this.init(element, options);
          }
}(window.jQuery);
```
 
Any help appreciated
Thanks a lot
Fabrizio

[bootstrap-checkbox-radio-switch.txt](https://github.com/turbolinks/turbolinks/files/1001824/bootstrap-checkbox-radio-switch.txt)

**Comment by domchristie** [comment]
>  I do not see any $(document).ready() statement or similar.

On line 100 and 242, there is the "shorthand" version of `$(document).ready()`:

```js
$(function () { … })
```

On line 100:

```js
$(function () {
  $('[data-toggle="checkbox"]').each(function () {
    var $checkbox = $(this);
    $checkbox.checkbox();
  });
});
```

and on 242:

```js
$(function () {
  $('[data-toggle="radio"]').each(function () {
    var $radio = $(this);
    $radio.radio();
  });
});
```

Duplicating this setup in a `turbolinks:load` event should hopefully fix your problem. For example, include the following in your main JavaScript bundle:

```js
$(document).on('turbolinks:load', function () {
  $('[data-toggle="checkbox"]').each(function () {
    var $checkbox = $(this);
    $checkbox.checkbox();
  });

  $('[data-toggle="radio"]').each(function () {
    var $radio = $(this);
    $radio.radio();
  });
})
```

The plugin seems to make sure that calling the the plugin more than once on the same element doesn't duplicate the behaviour (Lines 57-72 and 201-216), so hopefully there shouldn't be any side-effects from the library calling it on `$(document).ready`, and you calling it on `turbolinks:load`.

**Comment by fabriziobertoglio1987** [comment]
Thanks a lot !
Amazing this solved the issue. I had also some issues with bootstrap tooltips, but I fixed them the following post. I just write this as a note in case somebody in the future will have similar issues.

[(http://stackoverflow.com/questions/20129970/bootstrap-tooltip-in-ruby-on-rails#20130191)]

Your advice was very helpful, I will look deeper in this code in the future to make it more compatible with Turbolinks 5.

Unluckly I do not have much JS Experience, but now I will learn.

Thanks a lot
Fabrizio Bertoglio

**Comment by domchristie** [comment]
@fabriziobertoglio1987 can this issue be closed now?

**Comment by fabriziobertoglio1987** [comment]
@domchristie yes. thanks a lot. Sorry. I am experiencing problems getting github notifications.

**Comment by domchristie** [comment]
> Still it is not clear to me why many libraries break with Turbolinks 5 and where to find information on how to fix this problems.

Some libraries assume that pages are loaded in the _traditional_ way rather than via JavaScript. This means they are designed to run when the page is initially loaded. Turbolinks loads pages through JavaScript and so in order to use the library's code, you will often need to call their functions manually on `turbolinks:load`.

It's also worth noting that libraries often expect pages to be fully _unloaded_. This is not the case with Turbolinks as it caches the state of the page as you left it. So to prevent your library code from running more than once and causing duplication issues, it is often a good idea to teardown (or `destroy`) the effects of the plugin on `turbolinks:before-cache`.

So to summarise, if you must use plugins, looks for those that you can manually initialise as well as teardown. Initialise on `turbolinks:load` and destroy on `turbolinks:before-cache`.

If you have further issues, ask question on StackOverflow and tag it with 'turbolinks'. Hope that helps :)
